crates/anvil/src/eth/backend/mem/mod.rs

```diff
@@ -426,45 +426,59 @@ impl Backend {
+                    let mut env = self.env.read().clone();
+                    let (forked_db, client_fork_config) =
+                        node_config.setup_fork_db_config(fork_url, &mut env, &self.fees).await;
 
-                // this is the base fee of the current block, but we need the base fee of
-                // the next block
-                let next_block_base_fee = self.fees.get_next_block_base_fee_per_gas(
-                    fork_block.header.gas_used,
-                    fork_block.header.gas_limit,
-                    fork_block.header.base_fee_per_gas.unwrap_or_default(),
-                );
+                    *self.db.write().await = Box::new(forked_db);
+                    let fork = ClientFork::new(client_fork_config, Arc::clone(&self.db));
+                    *self.fork.write() = Some(fork);
+                    *self.env.write() = env;
+                } else {
+                    let mut env = self.env.write();
+                    env.cfg.chain_id = fork.chain_id();
+                    env.block = BlockEnv {
+                        number: U256::from(fork_block_number),
+                        timestamp: U256::from(fork_block.header.timestamp),
+                        gas_limit: U256::from(fork_block.header.gas_limit),
+                        difficulty: fork_block.header.difficulty,
+                        prevrandao: Some(fork_block.header.mix_hash.unwrap_or_default()),
+                        // Keep previous `coinbase` and `basefee` value
+                        coinbase: env.block.coinbase,
+                        basefee: env.block.basefee,
+                        ..env.block.clone()
+                    };
 
-                self.fees.set_base_fee(next_block_base_fee);
+                    // this is the base fee of the current block, but we need the base fee of
+                    // the next block
+                    let next_block_base_fee = self.fees.get_next_block_base_fee_per_gas(
+                        fork_block.header.gas_used,
+                        fork_block.header.gas_limit,
+                        fork_block.header.base_fee_per_gas.unwrap_or_default(),
+                    );
+
+                    self.fees.set_base_fee(next_block_base_fee);
+                }
+
+                // reset the time to the timestamp of the forked block
+                self.time.reset(fork_block.header.timestamp);
 
                 // also reset the total difficulty
                 self.blockchain.storage.write().total_difficulty = fork.total_difficulty();
             }
-
             // reset storage
             *self.blockchain.storage.write() = BlockchainStorage::forked(
                 fork.block_number(),
                 fork.block_hash(),
                 fork.total_difficulty(),
             );
             self.states.write().clear();
-            self.db.write().await.clear();
+            // self.db.write().await.clear();
```

Comment:
why do we no longer need this?

Reply:
Correct me if I'm wrong but we no longer need this because when we reset to a new forked state from a non-fork/old-fork state, we use `self.db`(`Backend.db`) as an intermediate step to set the `database` in `ClientFork`, [here](https://github.com/foundry-rs/foundry/blob/e613c1ab79f297ff46607fecf3066a0c958fc44c/crates/anvil/src/eth/backend/mem/mod.rs#L403-L405) and [here](https://github.com/foundry-rs/foundry/blob/e613c1ab79f297ff46607fecf3066a0c958fc44c/crates/anvil/src/eth/backend/mem/mod.rs#L438). 

So even when we clear - `self.db.write().await.clear();`, RPC calls work because they refer to the database in the fork + `foundry_fork_db::SharedBackend`. 

Ideally we should maintain this db in the `Backend` until we support resetting from forked state to non-forked (genesis) state.